### PR TITLE
SONARPY-1310 Add support of possessive quantifiers in regex parser for Python

### PR DESCRIPTION
**Pull request by maksim-grebeniuk-sonarsource**

### CODE REVIEW

**Review by guillaume-dequenne-sonarsource — CHANGES_REQUESTED** [review]
Looks good, some minor recommendations for test improvement.
Also, can you confirm the features work properly on the Python analyzer side before merging this PR?

File: regex-parsing/src/test/java/org/sonarsource/analyzer/commons/regex/python/PythonRegexSourceTest.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+import java.util.Set;
+import org.junit.jupiter.api.Test;
+import org.sonarsource.analyzer.commons.regex.CharacterParser;
+import org.sonarsource.analyzer.commons.regex.RegexFeature;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class PythonRegexSourceTest {
```
[the remark this answers is not in the document]
Comment (maksim-grebeniuk-sonarsource) [reply regex-parsing/src/test/java/org/sonarsource/analyzer/commons/regex/python/PythonRegexSourceTest.java]: done

File: regex-parsing/src/test/java/org/sonarsource/analyzer/commons/regex/python/PythonRegexSourceTest.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    var source = new PythonRegexSource("") {
+      @Override
+      public CharacterParser createCharacterParser() {
+        return null;
+      }
+    };
+
+    var features = source.features();
+    assertEquals(features.size(), 7);
+    assertTrue(features.containsAll(
```
[the remark this answers is not in the document]
Comment (maksim-grebeniuk-sonarsource) [reply regex-parsing/src/test/java/org/sonarsource/analyzer/commons/regex/python/PythonRegexSourceTest.java]: done

File: regex-parsing/src/test/java/org/sonarsource/analyzer/commons/regex/python/PythonRegexSourceTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  void testFeatures() {
+    var source = new PythonRegexSource("") {
+      @Override
+      public CharacterParser createCharacterParser() {
+        return null;
+      }
+    };
+
+    var features = source.features();
+    assertEquals(features.size(), 7);
```
[the remark this answers is not in the document]
Comment (maksim-grebeniuk-sonarsource) [reply regex-parsing/src/test/java/org/sonarsource/analyzer/commons/regex/python/PythonRegexSourceTest.java]: done

File: regex-parsing/src/test/java/org/sonarsource/analyzer/commons/regex/python/PythonRegexSourceTest.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    var source = new PythonRegexSource("") {
+      @Override
+      public CharacterParser createCharacterParser() {
+        return null;
+      }
+    };
+
+    var features = source.features();
+    Assertions.assertThat(features).hasSize(7);
+    Assertions.assertThat(features).containsOnly(
```
Comment (guillaume-dequenne-sonarsource) [inline regex-parsing/src/test/java/org/sonarsource/analyzer/commons/regex/python/PythonRegexSourceTest.java:40]: You can probably update the import to `import static org.assertj.core.api.Assertions.assertThat;` and chain the two assertions to make this a bit more concise.

**Pull request merged**